File: ExampleObjC/GiniVisionExampleObjC/ScreenAPIViewController.m
//
//  ViewController.m
//  GiniVisionExampleObjC
//
//  Created by Peter Pult on 21/06/16.
//  Copyright © 2016 Gini. All rights reserved.
//

#import "ScreenAPIViewController.h"
#import "AnalysisManager.h"
#import "ResultTableViewController.h"
#import "NoResultViewController.h"
#import <GiniVision/GiniVision-Swift.h>

@interface ScreenAPIViewController () <GiniVisionDelegate> {
    id<AnalysisDelegate> _analysisDelegate;
    NSData *_imageData;
}

@property (nonatomic, strong) NSString *errorMessage;
@property (nonatomic, strong) NSDictionary *result;
@property (nonatomic, strong) GINIDocument *document;

@end

@implementation ScreenAPIViewController

// MARK: View life cycle
- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self.navigationController setNavigationBarHidden:YES animated:animated];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self.navigationController setNavigationBarHidden:NO animated:animated];
}

// MARK: User actions
- (IBAction)easyLaunchGiniVision:(id)sender {
    
    /************************************************************************
     * CAPTURE IMAGE WITH THE SCREEN API OF THE GINI VISION LIBRARY FOR IOS *
     ************************************************************************/
    
    // 1. Create a custom configuration object
    GiniConfiguration *giniConfiguration = [GiniConfiguration new];
    giniConfiguration.debugModeOn = YES;
    giniConfiguration.navigationBarItemTintColor = [UIColor whiteColor];
    
    // 2. Create the Gini Vision Library view controller, set a delegate object and pass in the configuration object
    UIViewController *vc = [GiniVision viewControllerWithDelegate:self withConfiguration:giniConfiguration];
    
    // 3. Present the Gini Vision Library Screen API modally
    [self presentViewController:vc animated:YES completion:nil];
    
    // 4. Handle callbacks send out via the `GINIVisionDelegate` to get results, errors or updates on other user actions
}

// MARK: Gini Vision delegate
- (void)didCapture:(NSData *)imageData {
    NSLog(@"Screen API received image data");
    
    // Analyze image data right away with the Gini SDK for iOS to have results in as early as possible.
    [self analyzeDocumentWithImageData:imageData];
}

- (void)didReview:(NSData *)imageData withChanges:(BOOL)changes {
    NSString *changesString = changes ? @"changes" : @"no changes";
    NSLog(@"Screen API received updated image data with %@", changesString);
    
    // Analyze reviewed data because changes were made by the user during review.
    if (changes) {
        [self analyzeDocumentWithImageData:imageData];
        return;
    }
    
    // Present already existing results retrieved from the first analysis process initiated in `didCapture:`.
    if (_result && _document) {
        [self presentResults];
        return;
    }
    
    // Restart analysis if it was canceled and is currently not running.
    if (![[AnalysisManager sharedManager] isAnalyzing]) {
        [self analyzeDocumentWithImageData:imageData];
    }
}

- (void)didCancelCapturing {
    NSLog(@"Screen API canceled capturing");
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)didCancelReview {
    NSLog(@"Screen API canceled review");
    
    // Cancel analysis process to avoid unnecessary network calls.
    [self cancelAnalysis];
}

- (void)didShowAnalysis:(id<AnalysisDelegate>)analysisDelegate {
    NSLog(@"Screen API started analysis screen");
    
    _analysisDelegate = analysisDelegate;
    
    // The analysis screen is where the user should be confronted with any errors occuring during the analysis process.
    // Show any errors that occured while the user was still reviewing the image here.
    // Make sure to only show errors relevant to the user.
    [self showErrorMessage];
}

- (void)didCancelAnalysis {
    NSLog(@"Screen API canceled analysis");
    
    _analysisDelegate = nil;
    
    // Cancel analysis process to avoid unnecessary network calls.
    [self cancelAnalysis];
}

// MARK: Handle analysis of document
- (void)analyzeDocumentWithImageData:(NSData *)data {
    [self cancelAnalysis];
    _imageData = data;
    [[AnalysisManager sharedManager] analyzeDocumentWithImageData:data cancelationToken:[CancelationToken new] andCompletion:^(NSDictionary *result, GINIDocument * document, NSError *error) {
        if (error) {
            self.errorMessage = @"Es ist ein Fehler aufgetreten. Wiederholen";
        } else if (result && document) {
            self.document = document;
            self.result = result;
        } else {
            self.errorMessage = @"Ein unbekannter Fehler ist aufgetreten. Wiederholen";
        }
    }];
}

- (void)cancelAnalysis {
    [[AnalysisManager sharedManager] cancelAnalysis];
    _result = nil;
    _document = nil;
    _errorMessage = nil;
    _imageData = nil;
}

// MARK: Handle results from analysis process
- (void)setErrorMessage:(NSString *)errorMessage {
    _errorMessage = errorMessage;
    if (_errorMessage) {
        [self showErrorMessage];
    }
}

- (void)setResult:(NSDictionary *)result {
    _result = result;
    if (_result && _document) {
        [self showResults];
    }
}

- (void)showErrorMessage {
    if (_errorMessage && _imageData && _analysisDelegate) {
        [_analysisDelegate displayErrorWithMessage:_errorMessage andAction:^{
            [self analyzeDocumentWithImageData: _imageData];
        }];
    }
}

- (void)showResults {
    if (_analysisDelegate) {
        _analysisDelegate = nil;
        [self presentResults];
    }
}

- (void)presentResults {
    NSArray *payFive = @[@"paymentReference", @"iban", @"bic", @"amountToPay", @"paymentRecipient"];
    BOOL hasPayFive = NO;
    for (NSString *key in payFive) {
        if (_result[key]) {
            hasPayFive = YES;
            break;
        }
    }
    
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:NULL];
    if (hasPayFive) {
        ResultTableViewController *vc = [storyboard instantiateViewControllerWithIdentifier:@"resultScreen"];
        vc.result = _result;
        vc.document = _document;
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.navigationController pushViewController:vc animated:NO];
        });
    } else {
        NoResultViewController *vc = [storyboard instantiateViewControllerWithIdentifier:@"noResultScreen"];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.navigationController pushViewController:vc animated:NO];
        });
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self dismissViewControllerAnimated:YES completion:nil];
    });
}

@end

File: Example/GiniVision/ScreenAPIViewController.swift
//
//  ScreenAPIViewController.swift
//  GiniVision
//
//  Created by Peter Pult on 05/30/2016.
//  Copyright © 2016 Gini. All rights reserved.
//

import UIKit
import GiniVision
import Gini_iOS_SDK

/**
 View controller showing how to capture an image of a document using the Screen API of the Gini Vision Library for iOS
 and how to process it using the Gini SDK for iOS.
 */
class ScreenAPIViewController: UIViewController {
    
    @IBOutlet weak var metaInformationLabel: UILabel!
    
    var isUITest: Bool {
        return ProcessInfo.processInfo.arguments.contains("--UITest")
    }
    
    var analysisDelegate: GINIAnalysisDelegate?
    var imageData: Data?
    var result: GINIResult? {
        didSet {
            if let result = result,
               let document = document {
                show(result, fromDocument: document)
            }
        }
    }
    var document: GINIDocument?
    var errorMessage: String? {
        didSet {
            if let errorMessage = errorMessage {
                show(errorMessage: errorMessage)
            }
        }
    }
    
    // MARK: View life cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let customClientId = UserDefaults.standard.string(forKey: kSettingsGiniSDKClientIdKey) ?? ""
        let clientId = customClientId != "" ? customClientId : kGiniClientId
        
        metaInformationLabel.text = "Gini Vision Library: (\(GINIVision.versionString)) / Client id: \(clientId)"
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        self.navigationController?.setNavigationBarHidden(true, animated: animated)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        self.navigationController?.setNavigationBarHidden(false, animated: animated)
    }
    
    // MARK: User interaction
    @IBAction func easyLaunchGiniVision(_ sender: AnyObject) {
        
        /************************************************************************
         * CAPTURE IMAGE WITH THE SCREEN API OF THE GINI VISION LIBRARY FOR IOS *
         ************************************************************************/
        
        // 1. Create a custom configuration object
        let giniConfiguration = GINIConfiguration()
        giniConfiguration.debugModeOn = true
        giniConfiguration.navigationBarItemTintColor = UIColor.white
        
        // Make sure the app always behaves the same when run from UITests
        if isUITest {
            giniConfiguration.onboardingShowAtFirstLaunch = false
        }
        
        // 2. Create the Gini Vision Library view controller, set a delegate object and pass in the configuration object
        let vc = GINIVision.viewController(withDelegate: self, withConfiguration: giniConfiguration)
        
        // 3. Present the Gini Vision Library Screen API modally
        present(vc, animated: true, completion: nil)
        
        // 4. Handle callbacks send out via the `GINIVisionDelegate` to get results, errors or updates on other user actions
    }
    
    // MARK: Handle analysis of document
    func analyzeDocument(withImageData data: Data) {
        
        // Do not perform network tasks when UI testing.
        if isUITest {
            return
        }
        
        cancelAnalsyis()
        imageData = data
        AnalysisManager.sharedManager.analyzeDocument(withImageData: data, cancelationToken: CancelationToken(), completion: { inner in
            do {
                guard let response = try inner?(),
                      let result = response.0,
                      let document = response.1 else {
                        return self.errorMessage = "Ein unbekannter Fehler ist aufgetreten. Wiederholen"
                }
                self.document = document
                self.result = result
            } catch _ {
                self.errorMessage = "Es ist ein Fehler aufgetreten. Wiederholen"
            }
        })
    }
    
    func cancelAnalsyis() {
        AnalysisManager.sharedManager.cancelAnalysis()
        result = nil
        document = nil
        errorMessage = nil
        imageData = nil
    }
    
    // MARK: Handle results from analysis process
    func show(errorMessage message: String) {
        guard let imageData = self.imageData else {
            return
        }
        
        // Display an error with a custom message and custom action on the analysis screen
        analysisDelegate?.displayError(withMessage: errorMessage, andAction: {
            self.analyzeDocument(withImageData: imageData)
        })
    }
    
    func show(_ result: GINIResult, fromDocument document: GINIDocument) {
        if let _ = analysisDelegate {
            analysisDelegate = nil
            present(result, fromDocument: document)
        }
    }
    
    func present(_ result: GINIResult, fromDocument document: GINIDocument) {
        let payFive = ["paymentReference", "iban", "bic", "paymentReference", "amountToPay"]
        let hasPayFive = result.filter { payFive.contains($0.0) }.count > 0
        
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        if hasPayFive {
            let vc = storyboard.instantiateViewController(withIdentifier: "resultScreen") as! ResultTableViewController
            vc.result = result
            vc.document = document
            DispatchQueue.main.async {
                self.navigationController?.pushViewController(vc, animated: false)
            }
        } else {
            let vc = storyboard.instantiateViewController(withIdentifier: "noResultScreen") as! NoResultViewController
            DispatchQueue.main.async {
                self.navigationController?.pushViewController(vc, animated: false)
            }
        }
        
        DispatchQueue.main.async { 
            self.dismiss(animated: true, completion: nil)
        }
    }
}

// MARK: Gini Vision delegate
extension ScreenAPIViewController: GINIVisionDelegate {
    
    func didCapture(_ imageData: Data) {
        print("Screen API received image data")
        
        // Analyze image data right away with the Gini SDK for iOS to have results in as early as possible.
        analyzeDocument(withImageData: imageData)
    }
    
    func didReview(_ imageData: Data, withChanges changes: Bool) {
        print("Screen API received updated image data with \(changes ? "changes" : "no changes")")
        
        // Analyze reviewed data because changes were made by the user during review.
        if changes {
            analyzeDocument(withImageData: imageData)
            return
        }
        
        // Present already existing results retrieved from the first analysis process initiated in `didCapture:`.
        if let result = result,
           let document = document {
            present(result, fromDocument: document)
            return
        }
        
        // Restart analysis if it was canceled and is currently not running.
        if !AnalysisManager.sharedManager.isAnalyzing {
            analyzeDocument(withImageData: imageData)
        }
    }
    
    func didCancelCapturing() {
        print("Screen API canceled capturing")
        dismiss(animated: true, completion: nil)
    }
    
    // Optional delegate methods
    func didCancelReview() {
        print("Screen API canceled review")
        
        // Cancel analysis process to avoid unnecessary network calls.
        cancelAnalsyis()
    }
    
    func didShowAnalysis(_ analysisDelegate: GINIAnalysisDelegate) {
        print("Screen API started analysis screen")
        
        self.analysisDelegate = analysisDelegate
        
        // The analysis screen is where the user should be confronted with any errors occuring during the analysis process.
        // Show any errors that occured while the user was still reviewing the image here.
        // Make sure to only show errors relevant to the user.
        if let errorMessage = errorMessage {
            show(errorMessage: errorMessage)
        }
        
        // Send test error message when UI testing.
        if isUITest {
            analysisDelegate.displayError(withMessage: "My network error", andAction: { print("Try again") })
        }
    }
    
    func didCancelAnalysis() {
        print("Screen API canceled analysis")
        
        analysisDelegate = nil
        
        // Cancel analysis process to avoid unnecessary network calls.
        cancelAnalsyis()
    }
    
}
